components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/OIDCClaimUtil.java

```diff
@@ -270,4 +283,114 @@ private static boolean isUserConsentRequiredForClaims(String grantType) {
+
+    private static SubjectType getDefaultSubjectType() {
+
+        return StringUtils.isNotBlank(IdentityUtil.getProperty(DEFAULT_SUBJECT_TYPE)) ?
+                SubjectType.fromValue(IdentityUtil.getProperty(DEFAULT_SUBJECT_TYPE)) : SubjectType.PUBLIC;
+    }
+
+    /**
+     * Get the subject claim for the given user. If pairwise subject type is opted, then a PPID will be returned,
+     * otherwise, the authenticated user's username will be returned.
+     *
+     * @param clientId                       client id
+     * @param authenticatedSubjectIdentifier authenticated subject identifier
+     * @param callBackURI                    callback uri
+     * @return sub claim
+     * @throws IdentityOAuth2Exception when an error occurred while getting the service provider properties or user id
+     */
+    public static String getSubjectClaim(String clientId, String authenticatedSubjectIdentifier, String callBackURI)
+            throws IdentityOAuth2Exception {
+
+        if (SubjectType.PAIRWISE.equals(getSubjectType(clientId))) {
+            String sectorIdentifierUri = getSectorIdentifierUri(clientId);
+            return getPairwiseSubjectIdentifier(sectorIdentifierUri, authenticatedSubjectIdentifier, callBackURI);
+        }
+        return authenticatedSubjectIdentifier;
+    }
+
+    /**
+     * Get the sector identifier URI from the service provider properties.
+     *
+     * @param clientId client id
+     * @return sector identifier URI
+     * @throws IdentityOAuth2Exception when an error occurred while getting the service provider properties
+     */
+    private static String getSectorIdentifierUri(String clientId) throws IdentityOAuth2Exception {
+
+        ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+        ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+        for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+            if (SECTOR_IDENTIFIER_URI.equals(serviceProviderProperty.getName())) {
+                return serviceProviderProperty.getValue();
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Calculate pairwise subject identifier.
+     * <a href="https://openid.net/specs/openid-connect-core-1_0.html#PairwiseAlg">...</a>
+     *
+     * @param sectorIdentifierUri sector identifier URI
+     * @param userId              user id
+     * @param callBackURI         callback URI
+     * @return pairwise subject identifier
+     * @throws IdentityOAuth2Exception if required values are not present
+     */
+    private static String getPairwiseSubjectIdentifier(String sectorIdentifierUri, String userId, String callBackURI)
+            throws IdentityOAuth2Exception {
+
+        URI uri = StringUtils.isNotBlank(sectorIdentifierUri) ? URI.create(sectorIdentifierUri) :
+                StringUtils.isNotBlank(callBackURI) && isValidCallBackURI(callBackURI) ? URI.create(callBackURI) : null;
```

Comment:
Lets remove the ternary operators and add if conditions to make this more readable.